Load mod yaml via the virtual filesystem.

pchote opened the pull request:
This changes our rule/sequence/chrome/etc loaders to use the virtual filesystem code instead of .net filesystem code.  This is the next step towards adding support for oramod packages.

pchote committed:
Fix incorrect ModFiles.Open(resolved path).

pchote committed:
Load mod yaml via the virtual filesystem.

pchote committed:
Mount the mod filesystem context in the ModData ctor.

pchote commented:
It turns out that this change (and the plumbing we need for oramods in general) is fundamentally incompatible with the way our travis checks work.  Currently travis relies on not mounting anything in the filesystem as a workaround for not having a copy of the game assets.

We have three options:

1. Mark the downloaded content packages optional (using `~`) in the packages list, and rely on the load screen content checks to stop the game from loading without them.  This is a bit shit, but is the least invasive change.
2. Make travis download the mod content.  We can have travis cache the content files, which means this will have negligible effect on our infrastructure.  This would also let us remove hacks from the sequence parsers (which needs to defer any missing artwork errors to work around travis), and lets us add additional lint checks for sequence/sound references to missing assets. This is my preferred option because it makes our lint checks a lot more useful.  The downside is that mods that rely on proprietary assets that they can't distribute will not be able to use travis for lint checking.
3. Keep our current inconsistent mix of direct filesystem and vfs calls, and abandon our oramod plans (not realistic, but included for completeness).

Phrohdoh commented:
I think the obvious choice here for the engine is 2.
Thirdparty mods that want lint checking (or CI in general) are probably advanced enough to build their own system or find a solution that does work for them (local git hooks?).

This will bite us if/when RA2 is moved into OpenRA/OpenRA.

pchote commented:
>  The downside is that mods that rely on proprietary assets that they can't distribute will not be able to use travis for lint checking.

A solution for this would be for these mods to generate a dummy content package: files which match the structure and size of the proper files, but garbage content (e.g. fully transparent pixels for sprites, silence for sound files).

GraionDilach commented:
I guess I'll be participating in doing such assets for RA2, so size as frame size or bytesize.

pchote commented:
@GraionDilach best to have that discussion later in its own ticket if we decide to go with option 2.

pchote committed:
Make travis fetch the mod content before running tests.

pchote commented:
Added asset downloading and caching for Travis.

penev92 commented on OpenRA.Game/Scripting/ScriptContext.cs:
```diff
@@ -158,7 +159,7 @@ public sealed class ScriptContext : IDisposable
 
 			runtime.Globals["GameDir"] = Platform.GameDir;
-			runtime.DoBuffer(Game.ModData.ModFiles.Open(Platform.ResolvePath(".", "lua", "scriptwrapper.lua")).ReadAllText(), "scriptwrapper.lua").Dispose();
+			runtime.DoBuffer(File.Open(Platform.ResolvePath(".", "lua", "scriptwrapper.lua"), FileMode.Open).ReadAllText(), "scriptwrapper.lua").Dispose();
```
I thought we were moving away from the .NET/Mono file system in order to use our own?

pchote replied:
This is a core engine file (like the glsl shaders), so isn't referenced from the mod yaml.

penev92 commented on OpenRA.Mods.Common/UtilityCommands/ImportLegacyMapCommand.cs:
```diff
@@ -44,7 +44,6 @@ public virtual void Run(ModData modData, string[] args)
 		{
 			// HACK: The engine code assumes that Game.modData is set.
```
So when do we remove those?

pchote replied:
When the engine code stops assuming Game.modData is set :(

pchote replied:
There's about 200 more uses to remove first.

penev92 commented:
Fine :+1:

Mailaender commented:
You can enable `--check-sequence-sprites` checks now.

pchote commented:
Removed the two levels of laziness that used to hide sequence-asset mismatches from travis.  The third level of lazyness (at the Ruleset leve) remains, so this shouldn't affect the ingame perf.  Also removed the now-unnecessary `--check-sequence-sprites` utility command: sequence errors are now caught by the existing map rules lint pass.

Mailaender commented:
Looks like you forgot about mounting map assets like https://github.com/OpenRA/OpenRA/tree/bleed/mods/ra/maps/bomber-john

Mailaender commented:
Depends on https://github.com/OpenRA/ra2/pull/108 which currently is waiting for Travis admins to whitelist 7z for me.

pchote commented:
> Looks like you forgot about mounting map assets like https://github.com/OpenRA/OpenRA/tree/bleed/mods/ra/maps/bomber-john

Indeed I did.  Fixing this is non-trivial (we need to be able to mount the selected map in the lobby into its own isolated filesystem before we can validate its rules) and well outside the scope of this PR, so I have removed the commit that enables sequence linting and introduces the problem.

penev92 commented:
This PR *can't* possibly depend on a downstream fix for a random thirdparty mod. That's not how this works.

Mailaender commented on .travis.yml:
```diff
@@ -10,6 +10,7 @@ sudo: false
 cache:
   directories:
   - thirdparty/download
+  - Support
```
You probably should only cache `Support/Content`.

pchote commented:
Changed travis to only cache Support/Content.

Mailaender commented:
I think you can enable sprite sequence checks again. Without the refactoring commit `mono --debug OpenRA.Utility.exe ra --check-sequence-sprites` works fine on my machine using this pull request.

Mailaender commented:
Already came up with a solution for the Red Alert 2 CI checks. https://github.com/OpenRA/ra2/pull/108

pchote committed:
Lint sequences using the external utility command.

pchote commented:
Done.

obrakmann commented:
Custom maps don't seem to get loaded on this branch.

obrakmann commented:
:+1:

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/b286f1fd6d7b18e6508ba182134a1f5a5fceaa29)

pchote commented:
Turns out that this also broke dedicated servers and the resource site.  We probably want to go with option (1) in the short term so that we don't need a copy of the assets *everywhere*.
